File: clients/typescript/src/emoji.ts
// Emoji mapping utilities for Artissist Logger
import type { CustomEventMap, EmojiMapping, LogEvent } from './types';
import { TYPED_EMOJI_MAPPINGS } from './generated-types';

/**
 * Default emoji mappings for predefined events
 * Now uses the generated Smithy types for consistency across languages
 *
 * @deprecated Use TYPED_EMOJI_MAPPINGS from generated types for better type safety
 */
export const DEFAULT_EMOJI_MAPPINGS: Record<LogEvent, EmojiMapping> = TYPED_EMOJI_MAPPINGS;

/**
 * Legacy emoji mappings for backwards compatibility
 * Contains the old emoji mappings in case users need to reference them
 */
export const LEGACY_EMOJI_MAPPINGS: Record<LogEvent, EmojiMapping> = {
  SYSTEM_START: { emoji: '🚀', description: 'System startup or initialization', isDefault: true },
  SYSTEM_STOP: { emoji: '🛑', description: 'System shutdown or termination', isDefault: true },
  USER_AUTH: { emoji: '👤', description: 'User authentication events', isDefault: true },
  USER_AUTHZ: { emoji: '🔐', description: 'Authorization and permission events', isDefault: true },
  PROJECT_LIFECYCLE: { emoji: '📁', description: 'Project lifecycle events', isDefault: true },
  DATABASE_OPERATION: { emoji: '💾', description: 'Database operations', isDefault: true },
  API_REQUEST: { emoji: '🔄', description: 'API request/response events', isDefault: true },
  PERFORMANCE_METRIC: {
    emoji: '⚡',
    description: 'Performance metrics and timing',
    isDefault: true,
  },
  ERROR_OCCURRED: { emoji: '🐛', description: 'Error conditions and exceptions', isDefault: true },
  WARNING_ISSUED: { emoji: '⚠️', description: 'Warning conditions', isDefault: true },
  CONFIG_CHANGE: { emoji: '🔧', description: 'Configuration changes', isDefault: true },
  ANALYTICS_EVENT: { emoji: '📊', description: 'Analytics and tracking events', isDefault: true },
  AGENT_PROCESSING: { emoji: '🤖', description: 'Agent processing events', isDefault: true },
  CONVERSATION_EVENT: {
    emoji: '💬',
    description: 'Conversation and interaction events',
    isDefault: true,
  },
  ASSET_PROCESSING: { emoji: '📸', description: 'Asset upload and processing', isDefault: true },
  INSPIRATION_EVENT: { emoji: '🎨', description: 'Inspiration capture events', isDefault: true },
  INFRASTRUCTURE_DEPLOY: {
    emoji: '🏗️',
    description: 'Infrastructure deployment events',
    isDefault: true,
  },
  BUSINESS_METRIC: { emoji: '📈', description: 'Business metric events', isDefault: true },
  SEARCH_OPERATION: { emoji: '🔍', description: 'Search and discovery events', isDefault: true },
  // BREAKING CHANGE: These emojis have been standardized
  BACKGROUND_JOB: { emoji: '🔄', description: 'Background job processing', isDefault: true }, // was ⚙️
  NOTIFICATION_SENT: { emoji: '📧', description: 'Notification events', isDefault: true },
  SECURITY_EVENT: { emoji: '🔐', description: 'Security-related events', isDefault: true }, // was 🔒
  SCHEDULED_TASK: { emoji: '⏰', description: 'Scheduled task execution', isDefault: true },
  EXTERNAL_SERVICE: { emoji: '🌐', description: 'External service integration', isDefault: true },
  AUDIT_TRAIL: { emoji: '📋', description: 'Audit trail events', isDefault: true },
};

/**
 * Emoji resolver class for handling event-to-emoji mappings
 */
export class EmojiResolver {
  private customMappings: CustomEventMap = {};
  private enabled = false;

  constructor(enabled = false, customMappings: CustomEventMap = {}) {
    this.enabled = enabled;
    this.customMappings = customMappings;
  }

  /**
   * Enable or disable emoji resolution
   */
  setEnabled(enabled: boolean): void {
    this.enabled = enabled;
  }

  /**
   * Check if emoji resolution is enabled
   */
  isEnabled(): boolean {
    return this.enabled;
  }

  /**
   * Add or update custom event mappings
   */
  addCustomMapping(event: string, mapping: EmojiMapping): void {
    this.customMappings[event] = mapping;
  }

  /**
   * Add multiple custom event mappings
   */
  addCustomMappings(mappings: CustomEventMap): void {
    this.customMappings = { ...this.customMappings, ...mappings };
  }

  /**
   * Get emoji for a given event
   * @param event - The log event
   * @param fallbackEmoji - Optional fallback emoji if event not found
   * @returns The emoji string or empty string if disabled/not found
   */
  getEmoji(event: string, fallbackEmoji?: string): string {
    if (!this.enabled) {
      return '';
    }

    // Check custom mappings first
    if (typeof event === 'string' && this.customMappings[event]) {
      return this.customMappings[event].emoji;
    }

    // Check default mappings for LogEvent
    if (typeof event === 'string' && Object.keys(DEFAULT_EMOJI_MAPPINGS).includes(event)) {
      const logEvent = event as LogEvent;
      return DEFAULT_EMOJI_MAPPINGS[logEvent].emoji;
    }

    // Already checked as string above, so this is redundant
    // Type narrowing ensures we only deal with strings

    // Return fallback or empty string
    return fallbackEmoji ?? '';
  }

  /**
   * Get description for a given event
   */
  getDescription(event: string): string {
    // Check custom mappings first
    if (typeof event === 'string' && this.customMappings[event]) {
      return this.customMappings[event].description;
    }

    // Check default mappings
    if (typeof event === 'string' && Object.keys(DEFAULT_EMOJI_MAPPINGS).includes(event)) {
      const logEvent = event as LogEvent;
      return DEFAULT_EMOJI_MAPPINGS[logEvent].description;
    }

    // Already checked as string above, so this is redundant

    return 'Unknown event';
  }

  /**
   * Format a log message with emoji prefix
   */
  formatMessage(message: string, event?: string, fallbackEmoji?: string): string {
    if (!this.enabled || !event) {
      return message;
    }

    const emoji = this.getEmoji(event, fallbackEmoji);
    return emoji ? `${emoji} ${message}` : message;
  }

  /**
   * Get all available event mappings (default + custom)
   */
  getAllMappings(): Record<string, EmojiMapping> {
    return {
      ...DEFAULT_EMOJI_MAPPINGS,
      ...this.customMappings,
    };
  }

  /**
   * Validate emoji character (basic Unicode emoji check)
   */
  static isValidEmoji(emoji: string): boolean {
    // Basic emoji regex - checks for common emoji ranges
    const emojiRegex = /^[\u{1F300}-\u{1F9FF}\u{2600}-\u{26FF}\u{2700}-\u{27BF}]$/u;
    return emojiRegex.test(emoji);
  }

  /**
   * Create a new custom event mapping with validation
   */
  static createCustomMapping(emoji: string, description: string, isDefault = false): EmojiMapping {
    const trimmedEmoji = emoji.trim();

    if (!EmojiResolver.isValidEmoji(trimmedEmoji)) {
      throw new Error(`Invalid emoji character: ${trimmedEmoji}`);
    }

    if (!description || description.trim().length === 0) {
      throw new Error('Description cannot be empty');
    }

    return {
      emoji: trimmedEmoji,
      description: description.trim(),
      isDefault,
    };
  }
}

/**
 * Global emoji resolver instance
 * Can be used across the application for consistent emoji handling
 */
export const EventEmojiMapping = new EmojiResolver();

/**
 * Utility function to format log messages with emojis
 */
export function formatLogMessage(
  message: string,
  event?: string,
  enableEmojis = false,
  customResolver?: EmojiResolver
): string {
  const resolver = customResolver ?? EventEmojiMapping;

  if (!enableEmojis) {
    return message;
  }

  return resolver.formatMessage(message, event);
}

/**
 * Utility function to get all predefined events with their emojis
 */
export function getAllPredefinedEvents(): {
  event: LogEvent;
  emoji: string;
  description: string;
}[] {
  return Object.entries(DEFAULT_EMOJI_MAPPINGS).map(([event, mapping]) => ({
    event: event as LogEvent,
    emoji: mapping.emoji,
    description: mapping.description,
  }));
}

/**
 * Utility function to create a custom event registry
 */
export function createCustomEventRegistry(customMappings: CustomEventMap = {}): EmojiResolver {
  return new EmojiResolver(false, customMappings);
}

File: clients/python/artissist_logger/emoji.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""
Emoji resolution system for Artissist Logger Python client
"""

from dataclasses import dataclass
from typing import Dict, Optional

from .types import LogEvent
from .generated_types import TYPED_EMOJI_MAPPINGS


@dataclass
class EmojiMapping:
    """Mapping between events and emojis"""

    emoji: str
    description: str
    is_default: bool = True


class EmojiResolver:
    """Resolves log events to emoji representations"""

    # Legacy emoji mappings for backwards compatibility
    # Contains the old Python-specific emoji mappings
    LEGACY_MAPPINGS: Dict[LogEvent, EmojiMapping] = {
        LogEvent.SYSTEM_START: EmojiMapping(
            "🚀", "System startup or initialization"
        ),
        LogEvent.SYSTEM_STOP: EmojiMapping(
            "🛑", "System shutdown or termination"
        ),
        LogEvent.USER_AUTH: EmojiMapping("👤", "User authentication events"),
        LogEvent.USER_AUTHZ: EmojiMapping(
            "🔐", "Authorization and permission events"
        ),
        LogEvent.PROJECT_LIFECYCLE: EmojiMapping(
            "📁", "Project lifecycle events"
        ),
        LogEvent.DATABASE_OPERATION: EmojiMapping("💾", "Database operations"),
        LogEvent.API_REQUEST: EmojiMapping(
            "🔄", "API request/response events"
        ),
        LogEvent.PERFORMANCE_METRIC: EmojiMapping(
            "⚡", "Performance metrics and timing"
        ),
        LogEvent.ERROR_OCCURRED: EmojiMapping(
            "🐛", "Error conditions and exceptions"
        ),
        LogEvent.WARNING_ISSUED: EmojiMapping("⚠️", "Warning conditions"),
        LogEvent.CONFIG_CHANGE: EmojiMapping("🔧", "Configuration changes"),
        LogEvent.ANALYTICS_EVENT: EmojiMapping(
            "📊", "Analytics and tracking events"
        ),
        LogEvent.AGENT_PROCESSING: EmojiMapping(
            "🤖", "Agent processing events"
        ),
        LogEvent.CONVERSATION_EVENT: EmojiMapping(
            "💬", "Conversation and interaction events"
        ),
        # BREAKING CHANGE: These emojis have been unified with TypeScript
        LogEvent.ASSET_PROCESSING: EmojiMapping(
            "🖼️", "Asset upload and processing"  # was 🖼️, now 📸
        ),
        LogEvent.INSPIRATION_EVENT: EmojiMapping(
            "💡", "Inspiration capture events"  # was 💡, now 🎨
        ),
        LogEvent.INFRASTRUCTURE_DEPLOY: EmojiMapping(
            "🚢", "Infrastructure deployment events"  # was 🚢, now 🏗️
        ),
        LogEvent.BUSINESS_METRIC: EmojiMapping("📈", "Business metric events"),
        LogEvent.SEARCH_OPERATION: EmojiMapping(
            "🔍", "Search and discovery events"
        ),
        LogEvent.BACKGROUND_JOB: EmojiMapping(
            "⚙️", "Background job processing"
        ),
        LogEvent.NOTIFICATION_SENT: EmojiMapping(
            "📢", "Notification events"  # was 📢, now 📧
        ),
        LogEvent.SECURITY_EVENT: EmojiMapping("🔒", "Security-related events"),
        LogEvent.SCHEDULED_TASK: EmojiMapping(
            "⏰", "Scheduled task execution"
        ),
        LogEvent.EXTERNAL_SERVICE: EmojiMapping(
            "🔌", "External service integration"  # was 🔌, now 🌐
        ),
        LogEvent.AUDIT_TRAIL: EmojiMapping("📋", "Audit trail events"),
    }

    def __init__(
        self, custom_mappings: Optional[Dict[str, EmojiMapping]] = None
    ):
        """Initialize with optional custom emoji mappings"""
        self.custom_mappings = custom_mappings or {}
        self._default_mappings: Optional[Dict[LogEvent, EmojiMapping]] = None

    @property
    def default_mappings(self) -> Dict[LogEvent, EmojiMapping]:
        """Get default emoji mappings from generated Smithy types"""
        if self._default_mappings is None:
            self._default_mappings = {
                event: EmojiMapping(
                    str(config["emoji"]),
                    str(config["description"]),
                    bool(config["is_default"]),
                )
                for event, config in TYPED_EMOJI_MAPPINGS.items()
            }
        return self._default_mappings

    def get_emoji(
        self, event: Optional[LogEvent], custom_event: Optional[str] = None
    ) -> Optional[str]:
        """
        Get emoji for an event

        Args:
            event: Pre-defined LogEvent
            custom_event: Custom event name for extension

        Returns:
            Emoji string or None if not found
        """
        if event and event in self.default_mappings:
            return self.default_mappings[event].emoji

        if custom_event and custom_event in self.custom_mappings:
            return self.custom_mappings[custom_event].emoji

        return None

    def get_description(
        self, event: Optional[LogEvent], custom_event: Optional[str] = None
    ) -> Optional[str]:
        """
        Get description for an event

        Args:
            event: Pre-defined LogEvent
            custom_event: Custom event name for extension

        Returns:
            Description string or None if not found
        """
        if event and event in self.default_mappings:
            return self.default_mappings[event].description

        if custom_event and custom_event in self.custom_mappings:
            return self.custom_mappings[custom_event].description

        return None

    def add_custom_mapping(self, event_name: str, mapping: EmojiMapping):
        """Add or update a custom emoji mapping"""
        self.custom_mappings[event_name] = mapping

    def remove_custom_mapping(self, event_name: str) -> bool:
        """Remove a custom emoji mapping"""
        if event_name in self.custom_mappings:
            del self.custom_mappings[event_name]
            return True
        return False

    def list_all_mappings(self) -> Dict[str, EmojiMapping]:
        """Get all available emoji mappings"""
        result = {}

        # Add default mappings
        for event, mapping in self.default_mappings.items():
            result[event.value] = mapping

        # Add custom mappings
        result.update(self.custom_mappings)

        return result
